Fixup existing Makefile variables

permcody opened the pull request:
ADDITIONAL_LIBS is now added as a prereq to the target, this makes sense and should already be there.
EXTERNAL_FLAGS and ADDITIONAL_LIBS are switched in order in the final link stage. This is so that libraries
appeare before any external linker flags. This may or may not have adverse consequences.

closes #10308

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/136888/) on 07e1e64 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10309/site)

This comment will be updated on new commits.

andrsd commented:
magpie is using:
```
ADDITIONAL_LIBS += -L$(GSL_DIR) -lgsl -L$(GSL_DIR)/cblas -lgslcblas -I$(GSL_DIR)
```
to build some 3rd party packages, so that will throw `make` off. But using `EXTERNAL_FLAGS` should fix the problem.

Tagging @dschwen

permcody committed:
Adding a new dependency to the final application executable

Applications can add new libraries to the link line by defining them in $(ADDITIONAL_LIBS)
This breaks down a bit when absolute paths aren't provided or suffixes are omitted, which
is often the case in real target. By adding a new variable $(ADDITIONAL_DEPEND_LIBS) developers
can maintain full control of dependencies and proper link lines independently.

closes #10309

permcody commented:
Ping @andrsd, Can we get this in? It'll help with red crab builds.
